# Titanic competition with TensorFlow Decision Forests

This notebook will take you through the steps needed to train a baseline Gradient Boosted Trees Model using TensorFlow Decision Forests and creating a submission on the Titanic competition. 

This notebook shows:

1. How to do some basic pre-processing. For example, the passenger names will be tokenized, and ticket names will be splitted in parts.
1. How to train a Gradient Boosted Trees (GBT) with default parameters
1. How to train a GBT with improved default parameters
1. How to tune the parameters of a GBTs
1. How to train and ensemble many GBTs

# Imports dependencies

In [1]:
import numpy as np
import pandas as pd
import os

import tensorflow as tf
import tensorflow_decision_forests as tfdf

print(f"Found TF-DF {tfdf.__version__}")

Found TF-DF 1.2.0


# Load dataset

In [2]:
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
serving_df = pd.read_csv("/kaggle/input/titanic/test.csv")

train_df.head(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


# Prepare dataset

We will apply the following transformations on the dataset.

1. Tokenize the names. For example, "Braund, Mr. Owen Harris" will become ["Braund", "Mr.", "Owen", "Harris"].
2. Extract any prefix in the ticket. For example ticket "STON/O2. 3101282" will become "STON/O2." and 3101282.

In [3]:
def preprocess(df):
    df = df.copy()
    
    def normalize_name(x):
        return " ".join([v.strip(",()[].\"'") for v in x.split(" ")])
    
    def ticket_number(x):
        return x.split(" ")[-1]
        
    def ticket_item(x):
        items = x.split(" ")
        if len(items) == 1:
            return "NONE"
        return "_".join(items[0:-1])
    
    df["Name"] = df["Name"].apply(normalize_name)
    df["Ticket_number"] = df["Ticket"].apply(ticket_number)
    df["Ticket_item"] = df["Ticket"].apply(ticket_item)                     
    return df
    
preprocessed_train_df = preprocess(train_df)
preprocessed_serving_df = preprocess(serving_df)

preprocessed_train_df.head(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Ticket_number,Ticket_item
0,1,0,3,Braund Mr Owen Harris,male,22.0,1,0,A/5 21171,7.2500,NaN,S,21171,A/5
1,2,1,1,Cumings Mrs John Bradley Florence Briggs Thayer,female,38.0,1,0,PC 17599,71.2833,C85,C,17599,PC
2,3,1,3,Heikkinen Miss Laina,female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,3101282,STON/O2.
3,4,1,1,Futrelle Mrs Jacques Heath Lily May Peel,female,35.0,1,0,113803,53.1000,C123,S,113803,NONE
4,5,0,3,Allen Mr William Henry,male,35.0,0,0,373450,8.0500,NaN,S,373450,NONE


Let's keep the list of the input features of the model. Notably, we don't want to train our model on the "PassengerId" and "Ticket" features.

In [4]:
input_features = list(preprocessed_train_df.columns)
input_features.remove("Ticket")
input_features.remove("PassengerId")
input_features.remove("Survived")
#input_features.remove("Ticket_number")

print(f"Input features: {input_features}")

Input features: ['Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin', 'Embarked', 'Ticket_number', 'Ticket_item']


# Convert Pandas dataset to TensorFlow Dataset

In [5]:
def tokenize_names(features, labels=None):
    """Divite the names into tokens. TF-DF can consume text tokens natively."""
    features["Name"] =  tf.strings.split(features["Name"])
    return features, labels

train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_train_df,label="Survived").map(tokenize_names)
serving_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_serving_df).map(tokenize_names)

# Train model with default parameters

### Train model

First, we are training a GradientBoostedTreesModel model with the default parameters.

In [6]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    random_seed=1234,
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2026-03-25T17:00:42.207579966+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8e0bb97o/model/ with prefix aefabea040de4698
[INFO 2026-03-25T17:00:42.21625782+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-25T17:00:42.216388841+00:00 kernel.cc:1046] Use fast generic engine


Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: could not get source code
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Accuracy: 0.8260869383811951 Loss:0.8608942627906799


# Train model with improved default parameters

Now you'll use some specific parameters when creating the GBT model

In [7]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    
    #num_trees=2000,
    
    # Only for GBT.
    # A bit slower, but great to understand the model.
    # compute_permutation_variable_importance=True,
    
    # Change the default hyper-parameters
    # hyperparameter_template="benchmark_rank1@v1",
    
    #num_trees=1000,
    #tuner=tuner
    
    min_examples=1,
    categorical_algorithm="RANDOM",
    #max_depth=4,
    shrinkage=0.05,
    #num_candidate_attributes_ratio=0.2,
    split_axis="SPARSE_OBLIQUE",
    sparse_oblique_normalization="MIN_MAX",
    sparse_oblique_num_projections_exponent=2.0,
    num_trees=2000,
    #validation_ratio=0.0,
    random_seed=1234,
    
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2026-03-25T17:00:45.181434582+00:00 kernel.cc:1214] Loading model from path /tmp/tmpw6nwm5fa/model/ with prefix 0a0204b827e245bb
[INFO 2026-03-25T17:00:45.189914756+00:00 decision_forest.cc:661] Model loaded with 33 root(s), 1823 node(s), and 10 input feature(s).
[INFO 2026-03-25T17:00:45.189978806+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.760869562625885 Loss:1.0154211521148682


Let's look at the model and you can also notice the information about variable importance that the model figured out

In [8]:
model.summary()

Model: "gradient_boosted_trees_model_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
Total params: 1
Trainable params: 0
Non-trainable params: 1
_________________________________________________________________
Type: "GRADIENT_BOOSTED_TREES"
Task: CLASSIFICATION
Label: "__LABEL"

Input Features (11):
	Age
	Cabin
	Embarked
	Fare
	Name
	Parch
	Pclass
	Sex
	SibSp
	Ticket_item
	Ticket_number

No weights

Variable Importance: INV_MEAN_MIN_DEPTH:
    1.           "Sex"  0.576632 ################
    2.           "Age"  0.364297 #######
    3.          "Fare"  0.278839 ####
    4.          "Name"  0.208548 #
    5. "Ticket_number"  0.180792 
    6.        "Pclass"  0.176962 
    7.         "Parch"  0.176659 
    8.   "Ticket_item"  0.175540 
    9.      "Embarked"  0.172339 
   10.         "SibSp"  0.170442 

Variable Importance: NUM_AS_ROOT:
    1.  "Sex" 28.000000 ################
    2. "Name"  5.000000 

# Make predictions

In [9]:
def prediction_to_kaggle_format(model, threshold=0.5):
    proba_survive = model.predict(serving_ds, verbose=0)[:,0]
    return pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (proba_survive >= threshold).astype(int)
    })

def make_submission(kaggle_predictions):
    path="/kaggle/working/submission.csv"
    kaggle_predictions.to_csv(path, index=False)
    print(f"Submission exported to {path}")
    
kaggle_predictions = prediction_to_kaggle_format(model)
make_submission(kaggle_predictions)
!head /kaggle/working/submission.csv

Submission exported to /kaggle/working/submission.csv
PassengerId,Survived
892,0
893,0
894,0
895,0
896,0
897,0
898,0
899,0
900,1


# Training a model with hyperparameter tunning

Hyper-parameter tuning is enabled by specifying the tuner constructor argument of the model. The tuner object contains all the configuration of the tuner (search space, optimizer, trial and objective).


In [10]:
tuner = tfdf.tuner.RandomSearch(num_trials=1000)
tuner.choice("min_examples", [2, 5, 7, 10])
tuner.choice("categorical_algorithm", ["CART", "RANDOM"])

local_search_space = tuner.choice("growing_strategy", ["LOCAL"])
local_search_space.choice("max_depth", [3, 4, 5, 6, 8])

global_search_space = tuner.choice("growing_strategy", ["BEST_FIRST_GLOBAL"], merge=True)
global_search_space.choice("max_num_nodes", [16, 32, 64, 128, 256])

#tuner.choice("use_hessian_gain", [True, False])
tuner.choice("shrinkage", [0.02, 0.05, 0.10, 0.15])
tuner.choice("num_candidate_attributes_ratio", [0.2, 0.5, 0.9, 1.0])


tuner.choice("split_axis", ["AXIS_ALIGNED"])
oblique_space = tuner.choice("split_axis", ["SPARSE_OBLIQUE"], merge=True)
oblique_space.choice("sparse_oblique_normalization",
                     ["NONE", "STANDARD_DEVIATION", "MIN_MAX"])
oblique_space.choice("sparse_oblique_weights", ["BINARY", "CONTINUOUS"])
oblique_space.choice("sparse_oblique_num_projections_exponent", [1.0, 1.5])

# Tune the model. Notice the `tuner=tuner`.
tuned_model = tfdf.keras.GradientBoostedTreesModel(tuner=tuner)
tuned_model.fit(train_ds, verbose=0)

tuned_self_evaluation = tuned_model.make_inspector().evaluation()
print(f"Accuracy: {tuned_self_evaluation.accuracy} Loss:{tuned_self_evaluation.loss}")

Use /tmp/tmply0qp_fc as temporary training directory


[INFO 2026-03-25T17:02:08.043780647+00:00 kernel.cc:1214] Loading model from path /tmp/tmply0qp_fc/model/ with prefix b6a9b5393f544f13
[INFO 2026-03-25T17:02:08.053714937+00:00 decision_forest.cc:661] Model loaded with 19 root(s), 589 node(s), and 12 input feature(s).
[INFO 2026-03-25T17:02:08.053742367+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesGeneric" built
[INFO 2026-03-25T17:02:08.053758327+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.9178082346916199 Loss:0.6503586769104004


In the last line in the cell above, you can see the accuracy is higher than previously with default parameters and parameters set by hand.

This is the main idea behing hyperparameter tuning.

For more information you can follow this tutorial: [Automated hyper-parameter tuning](https://www.tensorflow.org/decision_forests/tutorials/automatic_tuning_colab)

# Making an ensemble

Here you'll create 100 models with different seeds and combine their results

This approach removes a little bit the random aspects related to creating ML models

In the GBT creation is used the `honest` parameter. It will use different training examples to infer the structure and the leaf values. This regularization technique trades examples for bias estimates.

In [11]:
predictions = None
num_predictions = 0

for i in range(100):
    print(f"i:{i}")
    # Possible models: GradientBoostedTreesModel or RandomForestModel
    model = tfdf.keras.GradientBoostedTreesModel(
        verbose=0, # Very few logs
        features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
        exclude_non_specified_features=True, # Only use the features in "features"

        #min_examples=1,
        #categorical_algorithm="RANDOM",
        ##max_depth=4,
        #shrinkage=0.05,
        ##num_candidate_attributes_ratio=0.2,
        #split_axis="SPARSE_OBLIQUE",
        #sparse_oblique_normalization="MIN_MAX",
        #sparse_oblique_num_projections_exponent=2.0,
        #num_trees=2000,
        ##validation_ratio=0.0,
        random_seed=i,
        honest=True,
    )
    model.fit(train_ds)
    
    sub_predictions = model.predict(serving_ds, verbose=0)[:,0]
    if predictions is None:
        predictions = sub_predictions
    else:
        predictions += sub_predictions
    num_predictions += 1

predictions/=num_predictions

kaggle_predictions = pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (predictions >= 0.5).astype(int)
    })

make_submission(kaggle_predictions)

i:0


[INFO 2026-03-25T17:02:08.802373702+00:00 kernel.cc:1214] Loading model from path /tmp/tmpuye3lmba/model/ with prefix 91bc4190dc144ee4
[INFO 2026-03-25T17:02:08.805692922+00:00 kernel.cc:1046] Use fast generic engine


i:1


[INFO 2026-03-25T17:02:09.958770439+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5epkjjjh/model/ with prefix 2b8a469d1a8e4e13
[INFO 2026-03-25T17:02:09.97262526+00:00 kernel.cc:1046] Use fast generic engine


i:2


[INFO 2026-03-25T17:02:10.729087085+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3afx2izo/model/ with prefix 01ea89498da1414f
[INFO 2026-03-25T17:02:10.732492975+00:00 kernel.cc:1046] Use fast generic engine


i:3


[INFO 2026-03-25T17:02:12.063326924+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_pmx6b6f/model/ with prefix 41e23f1a7ee04d40
[INFO 2026-03-25T17:02:12.082313201+00:00 kernel.cc:1046] Use fast generic engine


i:4


[INFO 2026-03-25T17:02:12.899404406+00:00 kernel.cc:1214] Loading model from path /tmp/tmpp7bi6ajk/model/ with prefix d970ca2913af413d
[INFO 2026-03-25T17:02:12.903502375+00:00 kernel.cc:1046] Use fast generic engine


i:5


[INFO 2026-03-25T17:02:13.711768051+00:00 kernel.cc:1214] Loading model from path /tmp/tmpup90wgw8/model/ with prefix b6fd437807d14463
[INFO 2026-03-25T17:02:13.714278381+00:00 kernel.cc:1046] Use fast generic engine


i:6


[INFO 2026-03-25T17:02:14.599386277+00:00 kernel.cc:1214] Loading model from path /tmp/tmplprt0xo0/model/ with prefix ecc493520ddc45e7
[INFO 2026-03-25T17:02:14.604984096+00:00 kernel.cc:1046] Use fast generic engine


i:7


[INFO 2026-03-25T17:02:15.8032824+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfwpih3qa/model/ with prefix dbdff012222f47c9
[INFO 2026-03-25T17:02:15.81833588+00:00 kernel.cc:1046] Use fast generic engine


i:8


[INFO 2026-03-25T17:02:16.753659912+00:00 kernel.cc:1214] Loading model from path /tmp/tmpirxmcg85/model/ with prefix 03efe81b69dc4963
[INFO 2026-03-25T17:02:16.760967381+00:00 kernel.cc:1046] Use fast generic engine


i:9


[INFO 2026-03-25T17:02:18.192756305+00:00 kernel.cc:1214] Loading model from path /tmp/tmprk4ydf9l/model/ with prefix e91647a193c14dc9
[INFO 2026-03-25T17:02:18.203959934+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-25T17:02:18.204000985+00:00 kernel.cc:1046] Use fast generic engine


i:10


[INFO 2026-03-25T17:02:19.027269722+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqgtw9mjw/model/ with prefix 0fe064c72100429b
[INFO 2026-03-25T17:02:19.031523901+00:00 kernel.cc:1046] Use fast generic engine


i:11


[INFO 2026-03-25T17:02:20.068079189+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvf_3bjod/model/ with prefix 46ee2280ace843ab
[INFO 2026-03-25T17:02:20.078769908+00:00 kernel.cc:1046] Use fast generic engine


i:12


[INFO 2026-03-25T17:02:20.897891509+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5gitsbma/model/ with prefix 8cdc0c9c62ff47e6
[INFO 2026-03-25T17:02:20.902303719+00:00 kernel.cc:1046] Use fast generic engine


i:13


[INFO 2026-03-25T17:02:21.840644222+00:00 kernel.cc:1214] Loading model from path /tmp/tmp719tzv_x/model/ with prefix 8f6f263a1af94b68
[INFO 2026-03-25T17:02:21.848994952+00:00 kernel.cc:1046] Use fast generic engine


i:14


[INFO 2026-03-25T17:02:22.663007221+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjvdecptc/model/ with prefix 574bfd5a94094c49
[INFO 2026-03-25T17:02:22.667520971+00:00 kernel.cc:1046] Use fast generic engine


i:15


[INFO 2026-03-25T17:02:23.51346674+00:00 kernel.cc:1214] Loading model from path /tmp/tmpep8j8z30/model/ with prefix 9fd03b0e1acf484a
[INFO 2026-03-25T17:02:23.51950857+00:00 kernel.cc:1046] Use fast generic engine


i:16


[INFO 2026-03-25T17:02:24.530219151+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmurqye0x/model/ with prefix 0e062056f3414e39
[INFO 2026-03-25T17:02:24.53953825+00:00 kernel.cc:1046] Use fast generic engine


i:17


[INFO 2026-03-25T17:02:25.598262625+00:00 kernel.cc:1214] Loading model from path /tmp/tmpghbuglh8/model/ with prefix 7ced99c4a6a349ef
[INFO 2026-03-25T17:02:25.608369664+00:00 kernel.cc:1046] Use fast generic engine


i:18


[INFO 2026-03-25T17:02:26.627583452+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwhnpu_qv/model/ with prefix b18fecae52ca4692
[INFO 2026-03-25T17:02:26.637142311+00:00 kernel.cc:1046] Use fast generic engine


i:19


[INFO 2026-03-25T17:02:27.834331479+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcc0m2az2/model/ with prefix e6871e5cdc3d4342
[INFO 2026-03-25T17:02:27.847902659+00:00 kernel.cc:1046] Use fast generic engine


i:20


[INFO 2026-03-25T17:02:28.933759499+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5qpm9zwm/model/ with prefix 66ce8dbbbb1f458a
[INFO 2026-03-25T17:02:28.944906268+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-25T17:02:28.944940598+00:00 kernel.cc:1046] Use fast generic engine


i:21


[INFO 2026-03-25T17:02:29.772758489+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3bu8d0hu/model/ with prefix a7d247314dff40f1
[INFO 2026-03-25T17:02:29.776703198+00:00 kernel.cc:1046] Use fast generic engine


i:22


[INFO 2026-03-25T17:02:30.58407533+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5fpntdy5/model/ with prefix 93a7eacd62bb40b6
[INFO 2026-03-25T17:02:30.588327609+00:00 kernel.cc:1046] Use fast generic engine


i:23


[INFO 2026-03-25T17:02:31.490045717+00:00 kernel.cc:1214] Loading model from path /tmp/tmprzgk6blb/model/ with prefix ed1f59e5f5c54a1b
[INFO 2026-03-25T17:02:31.496504097+00:00 kernel.cc:1046] Use fast generic engine


i:24


[INFO 2026-03-25T17:02:32.328444625+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9dxedxud/model/ with prefix c4ff1c2ee4e34def
[INFO 2026-03-25T17:02:32.332920075+00:00 kernel.cc:1046] Use fast generic engine


i:25


[INFO 2026-03-25T17:02:33.334465194+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvrl0ug2s/model/ with prefix 3df13854535643c6
[INFO 2026-03-25T17:02:33.343150903+00:00 kernel.cc:1046] Use fast generic engine


i:26


[INFO 2026-03-25T17:02:34.305373662+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0bzq6aej/model/ with prefix 4662bb47f37a473f
[INFO 2026-03-25T17:02:34.313079761+00:00 kernel.cc:1046] Use fast generic engine


i:27


[INFO 2026-03-25T17:02:35.188088232+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxxouxiws/model/ with prefix 9ef3e716dabd441b
[INFO 2026-03-25T17:02:35.193458361+00:00 kernel.cc:1046] Use fast generic engine


i:28


[INFO 2026-03-25T17:02:36.040361055+00:00 kernel.cc:1214] Loading model from path /tmp/tmphsmd_ze9/model/ with prefix e951b4ad16294260
[INFO 2026-03-25T17:02:36.044580234+00:00 kernel.cc:1046] Use fast generic engine


i:29


[INFO 2026-03-25T17:02:37.125022999+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxe_nd31d/model/ with prefix f46315179d634dcb
[INFO 2026-03-25T17:02:37.134982608+00:00 kernel.cc:1046] Use fast generic engine


i:30


[INFO 2026-03-25T17:02:38.620034051+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6tfk4g65/model/ with prefix 82d8913107cc4313
[INFO 2026-03-25T17:02:38.64017437+00:00 kernel.cc:1046] Use fast generic engine


i:31


[INFO 2026-03-25T17:02:40.195999821+00:00 kernel.cc:1214] Loading model from path /tmp/tmpi7ijl6t1/model/ with prefix 83b2876116374135
[INFO 2026-03-25T17:02:40.20547961+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-25T17:02:40.20552761+00:00 kernel.cc:1046] Use fast generic engine


i:32


[INFO 2026-03-25T17:02:41.11869504+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxsak13x4/model/ with prefix 4c7cf5102af14910
[INFO 2026-03-25T17:02:41.12339507+00:00 kernel.cc:1046] Use fast generic engine


i:33


[INFO 2026-03-25T17:02:42.160860495+00:00 kernel.cc:1214] Loading model from path /tmp/tmpokpdntnv/model/ with prefix 424aac5caa1f4b19
[INFO 2026-03-25T17:02:42.171342145+00:00 kernel.cc:1046] Use fast generic engine


i:34


[INFO 2026-03-25T17:02:43.095494985+00:00 kernel.cc:1214] Loading model from path /tmp/tmparf1ygfh/model/ with prefix 8343d3f8d02d4f2b
[INFO 2026-03-25T17:02:43.102243074+00:00 kernel.cc:1046] Use fast generic engine


i:35


[INFO 2026-03-25T17:02:44.049237563+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2vkefhuc/model/ with prefix 7bf3cae203564507
[INFO 2026-03-25T17:02:44.055372493+00:00 kernel.cc:1046] Use fast generic engine


i:36


[INFO 2026-03-25T17:02:45.129643246+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxvw08p8f/model/ with prefix 14df3ec475df4a48
[INFO 2026-03-25T17:02:45.140100536+00:00 kernel.cc:1046] Use fast generic engine


i:37


[INFO 2026-03-25T17:02:46.063858385+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkpz63bjp/model/ with prefix c44ddeaa85784915
[INFO 2026-03-25T17:02:46.070815685+00:00 kernel.cc:1046] Use fast generic engine


i:38


[INFO 2026-03-25T17:02:47.144988291+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3qpm29xt/model/ with prefix e6d268932ad94b18
[INFO 2026-03-25T17:02:47.155497481+00:00 kernel.cc:1046] Use fast generic engine


i:39


[INFO 2026-03-25T17:02:48.252390707+00:00 kernel.cc:1214] Loading model from path /tmp/tmpencjt2m8/model/ with prefix 842d76bdf6d94bae
[INFO 2026-03-25T17:02:48.262882905+00:00 kernel.cc:1046] Use fast generic engine


i:40


[INFO 2026-03-25T17:02:49.081969217+00:00 kernel.cc:1214] Loading model from path /tmp/tmpeofik6bm/model/ with prefix eaebd7479fa5473f
[INFO 2026-03-25T17:02:49.085318817+00:00 kernel.cc:1046] Use fast generic engine


i:41


[INFO 2026-03-25T17:02:50.222870227+00:00 kernel.cc:1214] Loading model from path /tmp/tmpm8dkvvvp/model/ with prefix 3991a7d2d5b64fb0
[INFO 2026-03-25T17:02:50.234791676+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-25T17:02:50.234829566+00:00 kernel.cc:1046] Use fast generic engine


i:42


[INFO 2026-03-25T17:02:51.240488963+00:00 kernel.cc:1214] Loading model from path /tmp/tmp34nd3_jh/model/ with prefix 57f724c681624f2b
[INFO 2026-03-25T17:02:51.247477842+00:00 kernel.cc:1046] Use fast generic engine


i:43


[INFO 2026-03-25T17:02:52.475302211+00:00 kernel.cc:1214] Loading model from path /tmp/tmpg27enmhw/model/ with prefix b591366ad7ad4802
[INFO 2026-03-25T17:02:52.48891552+00:00 kernel.cc:1046] Use fast generic engine


i:44


[INFO 2026-03-25T17:02:53.466081333+00:00 kernel.cc:1214] Loading model from path /tmp/tmphommyruy/model/ with prefix 9768044552d64e61
[INFO 2026-03-25T17:02:53.473641992+00:00 kernel.cc:1046] Use fast generic engine


i:45


[INFO 2026-03-25T17:02:54.280351438+00:00 kernel.cc:1214] Loading model from path /tmp/tmp19inxsfp/model/ with prefix 5334af33e3354df4
[INFO 2026-03-25T17:02:54.283136738+00:00 kernel.cc:1046] Use fast generic engine


i:46


[INFO 2026-03-25T17:02:55.477454743+00:00 kernel.cc:1214] Loading model from path /tmp/tmpprta8pv7/model/ with prefix c1ff78a540cc450f
[INFO 2026-03-25T17:02:55.488909732+00:00 kernel.cc:1046] Use fast generic engine


i:47


[INFO 2026-03-25T17:02:56.641724521+00:00 kernel.cc:1214] Loading model from path /tmp/tmpakjfugs7/model/ with prefix 7c296b5a92484e58
[INFO 2026-03-25T17:02:56.65241747+00:00 kernel.cc:1046] Use fast generic engine


i:48


[INFO 2026-03-25T17:02:57.529483938+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8erv2k94/model/ with prefix a52d7bb8d2ee4ff9
[INFO 2026-03-25T17:02:57.533036437+00:00 kernel.cc:1046] Use fast generic engine


i:49


[INFO 2026-03-25T17:02:58.505482621+00:00 kernel.cc:1214] Loading model from path /tmp/tmpw1nszy2q/model/ with prefix b4b9384199944c7e
[INFO 2026-03-25T17:02:58.510865211+00:00 kernel.cc:1046] Use fast generic engine


i:50


[INFO 2026-03-25T17:02:59.586153561+00:00 kernel.cc:1214] Loading model from path /tmp/tmpg5u54byt/model/ with prefix 162364c308e34e02
[INFO 2026-03-25T17:02:59.595699499+00:00 kernel.cc:1046] Use fast generic engine


i:51


[INFO 2026-03-25T17:03:00.790355234+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvb5qb0v1/model/ with prefix 7399fb490ef4430c
[INFO 2026-03-25T17:03:00.802490503+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-25T17:03:00.802528563+00:00 kernel.cc:1046] Use fast generic engine


i:52


[INFO 2026-03-25T17:03:01.759069675+00:00 kernel.cc:1214] Loading model from path /tmp/tmpph_tqr4v/model/ with prefix fdeb7a4515c44ff6
[INFO 2026-03-25T17:03:01.764994514+00:00 kernel.cc:1046] Use fast generic engine


i:53


[INFO 2026-03-25T17:03:02.664164388+00:00 kernel.cc:1214] Loading model from path /tmp/tmpublgeibq/model/ with prefix be8ef2c0213d45d7
[INFO 2026-03-25T17:03:02.670054767+00:00 kernel.cc:1046] Use fast generic engine


i:54


[INFO 2026-03-25T17:03:03.500603874+00:00 kernel.cc:1214] Loading model from path /tmp/tmpai0f7unp/model/ with prefix 4a5fc929b0be4b8d
[INFO 2026-03-25T17:03:03.503232534+00:00 kernel.cc:1046] Use fast generic engine


i:55


[INFO 2026-03-25T17:03:04.715838931+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfzze_tit/model/ with prefix 5408ff6e74c740c2
[INFO 2026-03-25T17:03:04.726793699+00:00 kernel.cc:1046] Use fast generic engine


i:56


[INFO 2026-03-25T17:03:06.498686986+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1bkpm5tz/model/ with prefix 99298f27daf84697
[INFO 2026-03-25T17:03:06.509409325+00:00 kernel.cc:1046] Use fast generic engine


i:57


[INFO 2026-03-25T17:03:07.485650809+00:00 kernel.cc:1214] Loading model from path /tmp/tmpy02o0a4b/model/ with prefix b1971624e2b24624
[INFO 2026-03-25T17:03:07.489238468+00:00 kernel.cc:1046] Use fast generic engine


i:58


[INFO 2026-03-25T17:03:08.477027986+00:00 kernel.cc:1214] Loading model from path /tmp/tmphqoexu9l/model/ with prefix 2a1ba193f4494fdf
[INFO 2026-03-25T17:03:08.482310756+00:00 kernel.cc:1046] Use fast generic engine


i:59


[INFO 2026-03-25T17:03:09.500248332+00:00 kernel.cc:1214] Loading model from path /tmp/tmpp62l633e/model/ with prefix 7010e650d5ba437e
[INFO 2026-03-25T17:03:09.507979321+00:00 kernel.cc:1046] Use fast generic engine


i:60


[INFO 2026-03-25T17:03:10.523395067+00:00 kernel.cc:1214] Loading model from path /tmp/tmp891xxhuw/model/ with prefix 59ea285012cd4859
[INFO 2026-03-25T17:03:10.531232566+00:00 kernel.cc:1046] Use fast generic engine


i:61


[INFO 2026-03-25T17:03:11.445054454+00:00 kernel.cc:1214] Loading model from path /tmp/tmpeg1_9n58/model/ with prefix 133cf8c0a5d145e1
[INFO 2026-03-25T17:03:11.449047293+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-25T17:03:11.449081553+00:00 kernel.cc:1046] Use fast generic engine


i:62


[INFO 2026-03-25T17:03:12.88453134+00:00 kernel.cc:1214] Loading model from path /tmp/tmpob21p0cw/model/ with prefix 627ec9b12a544f18
[INFO 2026-03-25T17:03:12.904249018+00:00 kernel.cc:1046] Use fast generic engine


i:63


[INFO 2026-03-25T17:03:13.876963386+00:00 kernel.cc:1214] Loading model from path /tmp/tmpu0j2p7hp/model/ with prefix 9d2bed3fa5ae4421
[INFO 2026-03-25T17:03:13.884501606+00:00 kernel.cc:1046] Use fast generic engine


i:64


[INFO 2026-03-25T17:03:14.883944752+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdtoi9hh_/model/ with prefix bd49c1d6077d46fe
[INFO 2026-03-25T17:03:14.889993171+00:00 kernel.cc:1046] Use fast generic engine


i:65


[INFO 2026-03-25T17:03:15.784585397+00:00 kernel.cc:1214] Loading model from path /tmp/tmpca0pjq95/model/ with prefix be629d2c9eb5434c
[INFO 2026-03-25T17:03:15.788432966+00:00 kernel.cc:1046] Use fast generic engine


i:66


[INFO 2026-03-25T17:03:16.737766615+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0jgwagm4/model/ with prefix 7d122fe5a254433f
[INFO 2026-03-25T17:03:16.742705605+00:00 kernel.cc:1046] Use fast generic engine


i:67


[INFO 2026-03-25T17:03:18.034190792+00:00 kernel.cc:1214] Loading model from path /tmp/tmp43sq8ny_/model/ with prefix 65d51d2fbfc54497
[INFO 2026-03-25T17:03:18.048848321+00:00 kernel.cc:1046] Use fast generic engine


i:68


[INFO 2026-03-25T17:03:19.161398973+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsfl4wsm_/model/ with prefix 7f68a1de90c54d3d
[INFO 2026-03-25T17:03:19.170784933+00:00 kernel.cc:1046] Use fast generic engine


i:69


[INFO 2026-03-25T17:03:20.122546866+00:00 kernel.cc:1214] Loading model from path /tmp/tmpne6vwrfg/model/ with prefix 09001b3118744ce2
[INFO 2026-03-25T17:03:20.127598426+00:00 kernel.cc:1046] Use fast generic engine


i:70


[INFO 2026-03-25T17:03:21.151124014+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3qyw0sbm/model/ with prefix d4935b5d088f4481
[INFO 2026-03-25T17:03:21.158043373+00:00 kernel.cc:1046] Use fast generic engine


i:71


[INFO 2026-03-25T17:03:22.14897867+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3n0llddi/model/ with prefix b7483d9cf81b48ae
[INFO 2026-03-25T17:03:22.154945949+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-25T17:03:22.154991479+00:00 kernel.cc:1046] Use fast generic engine


i:72


[INFO 2026-03-25T17:03:23.404715112+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2ifptao8/model/ with prefix 28b787008d31494b
[INFO 2026-03-25T17:03:23.416915961+00:00 kernel.cc:1046] Use fast generic engine


i:73


[INFO 2026-03-25T17:03:24.324229163+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_vkuz4ap/model/ with prefix c4f15de068a54fab
[INFO 2026-03-25T17:03:24.329585493+00:00 kernel.cc:1046] Use fast generic engine


i:74


[INFO 2026-03-25T17:03:25.41591386+00:00 kernel.cc:1214] Loading model from path /tmp/tmps75hfm07/model/ with prefix e8ee17294b204e0e
[INFO 2026-03-25T17:03:25.424645649+00:00 kernel.cc:1046] Use fast generic engine


i:75


[INFO 2026-03-25T17:03:26.384056367+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4ryenh2i/model/ with prefix 4fb2fd563c4a441d
[INFO 2026-03-25T17:03:26.390248996+00:00 kernel.cc:1046] Use fast generic engine


i:76


[INFO 2026-03-25T17:03:27.234995224+00:00 kernel.cc:1214] Loading model from path /tmp/tmpj1sbioij/model/ with prefix fb864ccec170485c
[INFO 2026-03-25T17:03:27.238244703+00:00 kernel.cc:1046] Use fast generic engine


i:77


[INFO 2026-03-25T17:03:28.08881516+00:00 kernel.cc:1214] Loading model from path /tmp/tmpf2cw_8y4/model/ with prefix 914ef2b9690d461e
[INFO 2026-03-25T17:03:28.09244815+00:00 kernel.cc:1046] Use fast generic engine


i:78


[INFO 2026-03-25T17:03:28.999194644+00:00 kernel.cc:1214] Loading model from path /tmp/tmp760961j4/model/ with prefix 3914803ff4e4437c
[INFO 2026-03-25T17:03:29.004390313+00:00 kernel.cc:1046] Use fast generic engine


i:79


[INFO 2026-03-25T17:03:29.941490139+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqf6i9ut6/model/ with prefix 1841bc3f316e4bdc
[INFO 2026-03-25T17:03:29.947640999+00:00 kernel.cc:1046] Use fast generic engine


i:80


[INFO 2026-03-25T17:03:30.972064376+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7cwd3218/model/ with prefix 9be668fa3cbd4c78
[INFO 2026-03-25T17:03:30.979771055+00:00 kernel.cc:1046] Use fast generic engine


i:81


[INFO 2026-03-25T17:03:32.042928509+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwn23yu4_/model/ with prefix 53d04c8a8d70464a
[INFO 2026-03-25T17:03:32.051867638+00:00 kernel.cc:1046] Use fast generic engine


i:82


[INFO 2026-03-25T17:03:33.066542+00:00 kernel.cc:1214] Loading model from path /tmp/tmpytv6svq6/model/ with prefix 44439d1134584a87
[INFO 2026-03-25T17:03:33.074597509+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-25T17:03:33.074637859+00:00 kernel.cc:1046] Use fast generic engine


i:83


[INFO 2026-03-25T17:03:34.079000512+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnic60t4h/model/ with prefix 46954b9f67aa4a86
[INFO 2026-03-25T17:03:34.085697041+00:00 kernel.cc:1046] Use fast generic engine


i:84


[INFO 2026-03-25T17:03:36.065086549+00:00 kernel.cc:1214] Loading model from path /tmp/tmpe2riyx50/model/ with prefix de96d8989a214654
[INFO 2026-03-25T17:03:36.079461678+00:00 kernel.cc:1046] Use fast generic engine


i:85


[INFO 2026-03-25T17:03:37.03955991+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9cjhzmhk/model/ with prefix 0c146e39e0c34987
[INFO 2026-03-25T17:03:37.04474722+00:00 kernel.cc:1046] Use fast generic engine


i:86


[INFO 2026-03-25T17:03:38.274853369+00:00 kernel.cc:1214] Loading model from path /tmp/tmpw4j2_4tm/model/ with prefix d5ef2c3f493b43a8
[INFO 2026-03-25T17:03:38.287541377+00:00 kernel.cc:1046] Use fast generic engine


i:87


[INFO 2026-03-25T17:03:39.559297679+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9t8t60gr/model/ with prefix cd954c7dfd9c446c
[INFO 2026-03-25T17:03:39.573285518+00:00 kernel.cc:1046] Use fast generic engine


i:88


[INFO 2026-03-25T17:03:40.621239988+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmj0x1738/model/ with prefix 47a5fd2b04e646df
[INFO 2026-03-25T17:03:40.630575226+00:00 kernel.cc:1046] Use fast generic engine


i:89


[INFO 2026-03-25T17:03:41.558023251+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfd5ohbxf/model/ with prefix 1d33111877354b77
[INFO 2026-03-25T17:03:41.561876751+00:00 kernel.cc:1046] Use fast generic engine


i:90


[INFO 2026-03-25T17:03:42.592213516+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1655azeh/model/ with prefix ab6a687c85b940de
[INFO 2026-03-25T17:03:42.599986426+00:00 kernel.cc:1046] Use fast generic engine


i:91


[INFO 2026-03-25T17:03:43.504535093+00:00 kernel.cc:1214] Loading model from path /tmp/tmpagpls81_/model/ with prefix 1aafae4a5f1a4fd8
[INFO 2026-03-25T17:03:43.509913992+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2026-03-25T17:03:43.509948272+00:00 kernel.cc:1046] Use fast generic engine


i:92


[INFO 2026-03-25T17:03:44.770195898+00:00 kernel.cc:1214] Loading model from path /tmp/tmpj8yd0ulu/model/ with prefix 1fec68e041c24244
[INFO 2026-03-25T17:03:44.783988876+00:00 kernel.cc:1046] Use fast generic engine


i:93


[INFO 2026-03-25T17:03:45.917773381+00:00 kernel.cc:1214] Loading model from path /tmp/tmppi56agz8/model/ with prefix 4253e9aa815d440b
[INFO 2026-03-25T17:03:45.92568987+00:00 kernel.cc:1046] Use fast generic engine


i:94


[INFO 2026-03-25T17:03:46.936925567+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9cnn7s4k/model/ with prefix 4480d08ddf0a47f2
[INFO 2026-03-25T17:03:46.942037707+00:00 kernel.cc:1046] Use fast generic engine


i:95


[INFO 2026-03-25T17:03:47.99978837+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbxoon9_e/model/ with prefix c030db313e964d12
[INFO 2026-03-25T17:03:48.006961169+00:00 kernel.cc:1046] Use fast generic engine


i:96


[INFO 2026-03-25T17:03:49.109219118+00:00 kernel.cc:1214] Loading model from path /tmp/tmpp0uq5hj_/model/ with prefix a580e0aba4c645e4
[INFO 2026-03-25T17:03:49.117279437+00:00 kernel.cc:1046] Use fast generic engine


i:97


[INFO 2026-03-25T17:03:50.031815778+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdlaxgc5v/model/ with prefix 377548ac9233462f
[INFO 2026-03-25T17:03:50.036058207+00:00 kernel.cc:1046] Use fast generic engine


i:98


[INFO 2026-03-25T17:03:51.10110282+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3we1kr8s/model/ with prefix 2ad5c67cd4a142ac
[INFO 2026-03-25T17:03:51.109106169+00:00 kernel.cc:1046] Use fast generic engine


i:99


[INFO 2026-03-25T17:03:52.376711011+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvf5w29ma/model/ with prefix bf154a31763040e5
[INFO 2026-03-25T17:03:52.38782642+00:00 kernel.cc:1046] Use fast generic engine


Submission exported to /kaggle/working/submission.csv


# What is next

If you want to learn more about TensorFlow Decision Forests and its advanced features, you can follow the official documentation [here](https://www.tensorflow.org/decision_forests) 